### PR TITLE
V5 keys

### PR DESCRIPTION
**Pull request by zugzwang**

### CODE REVIEW

File: openpgp/packet/signature.go [hunk 1]
```diff
@@ -706,11 +764,13 @@ func (sig *Signature) Serialize(w io.Writer) (err error) {
 	length := len(sig.HashSuffix) - 6 /* trailer not included */ +
 		2 /* length of unhashed subpackets */ + unhashedSubpacketsLen +
 		2 /* hash tag */ + sigLength
+	if sig.Version() == 5 {
+		length -= 4 // eight-octet instead of four-octet big endian
+	}
```
Comment (twiss) [inline openpgp/packet/signature.go:769]: And then also here, I think this is not correct for binary and text signatures (the v5-specific fields are included in the length but shouldn't be). So instead of starting with the length of `sig.HashSuffix` and then subtracting stuff, I would compute the length from the ground up here (i.e. a constant plus the length of the hashed subpackets, plus the unhashed subpackets as it's already doing).

File: openpgp/packet/signature.go [hunk 2]
```diff
@@ -720,7 +780,14 @@ func (sig *Signature) Serialize(w io.Writer) (err error) {
 
 func (sig *Signature) serializeBody(w io.Writer) (err error) {
 	unhashedSubpacketsLen := subpacketsLength(sig.outSubpackets, false)
-	_, err = w.Write(sig.HashSuffix[:len(sig.HashSuffix)-6])
+	fields := sig.HashSuffix[:len(sig.HashSuffix)-6]
+	if sig.Version() == 5 {
+		fields = fields[:len(fields)-4]
+		if sig.SigType == SigTypeBinary || sig.SigType == SigTypeText {
+			fields = fields[:len(fields)-6]
```
Comment (twiss) [inline openpgp/packet/signature.go:787]: And then also here, I would take the first `n` bytes (calculated in the same way as above) instead of everything except the last `m` bytes; or alternatively, just write the values directly (same as https://github.com/ProtonMail/crypto/blob/dbafc501ce7a096d3ff2d828c9638488c77c8cc4/openpgp/packet/signature.go#L556-L565).

File: openpgp/packet/signature.go [hunk 3]
```diff
@@ -774,14 +841,19 @@ type outputSubpacket struct {
...
 
-	if sig.IssuerKeyId != nil {
+	if sig.IssuerKeyId != nil && sig.Version() == 4 {
 		keyId := make([]byte, 8)
 		binary.BigEndian.PutUint64(keyId, *sig.IssuerKeyId)
-		subpackets = append(subpackets, outputSubpacket{true, issuerSubpacket, false, keyId})
+		subpackets = append(subpackets, outputSubpacket{true, issuerSubpacket, true, keyId})
+	}
+	if sig.IssuerKeyFingerprint != nil {
+		contents := append([]uint8{uint8(sig.Version())}, sig.IssuerKeyFingerprint...)
```
Comment (twiss) [inline openpgp/packet/signature.go:854]: Technically speaking, this should be the key version, not the signature version. Obviously, they are always the same right now, but I think technically speaking this is not required, so it might be nice to make it obvious in the code, e.g. by adding an explicit `IssuerKeyVersion *int` property?

File: openpgp/packet/signature.go [hunk 4]
```diff
@@ -774,14 +841,19 @@ type outputSubpacket struct {
 func (sig *Signature) buildSubpackets() (subpackets []outputSubpacket, err error) {
 	creationTime := make([]byte, 4)
 	binary.BigEndian.PutUint32(creationTime, uint32(sig.CreationTime.Unix()))
-	subpackets = append(subpackets, outputSubpacket{true, creationTimeSubpacket, false, creationTime})
+	subpackets = []outputSubpacket{
+		outputSubpacket{true, creationTimeSubpacket, false, creationTime},
+	}
```
Comment (twiss) [inline openpgp/packet/signature.go:846]: (Why this change?)

File: openpgp/packet/signature.go [hunk 5]
```diff
@@ -860,8 +935,7 @@ func (sig *Signature) buildSubpackets() (subpackets []outputSubpacket, err error
 		if err != nil {
 			return
 		}
-		subpackets = append(subpackets, outputSubpacket{true, embeddedSignatureSubpacket, true,
-			buf.Bytes()})
+		subpackets = append(subpackets, outputSubpacket{true, embeddedSignatureSubpacket, true, buf.Bytes()})
```
Comment (twiss) [inline openpgp/packet/signature.go:938]: Not that it matters too much, but I would try to limit the amount of trivial changes in this PR, because they make it more difficult to see which changes are important. Also for some whitespace changes above; I would just take a quick pass through the PR and remove everything that's not necessary.

File: openpgp/packet/signature.go [hunk 6]
```diff
@@ -491,6 +513,25 @@ func serializeSubpackets(to []byte, subpackets []outputSubpacket, hashed bool) {
...
+}
+
+// MatchIssuerVersion copies the version from the issuer public key
+// (implementations MUST generate version 5 signatures when using a version 5
+// key, and SHOULD generate V4 signatures with version 4 keys). If the key is
+// V5, it upgrades the signature with the necessary fields.
+func (sig *Signature) MatchIssuerVersion(issuer *PublicKey) {
+	sig.version = issuer.Version()
+	if sig.version == 5 {
+		sig.IssuerKeyFingerprint = issuer.Fingerprint
```
Comment (twiss) [inline openpgp/packet/signature.go:531]: I would always add the issuer fingerprint, regardless of version; see https://github.com/openpgpjs/openpgpjs/pull/1097. (And do so directly in `Signature.Sign`.)

File: openpgp/write.go [hunk 7]
```diff
@@ -68,6 +68,7 @@ func detachSign(w io.Writer, signer *Entity, message io.Reader, sigType packet.S
 	}
 
 	sig := new(packet.Signature)
+	sig.MatchIssuerVersion(&signer.PrivateKey.PublicKey)
```
Comment (twiss) [inline openpgp/write.go:71]: I think this one isn't needed, if we already do so in `Signature.Sign`? (And the same thing in key_generation.go.)

**Review by twiss — COMMENTED** [review]
Looks like a good start to me :+1: Some comments below.

File: openpgp/packet/public_key.go [hunk 8]
```diff
@@ -35,10 +35,11 @@ type kdfAlgorithm byte
 
 // PublicKey represents an OpenPGP public key. See RFC 4880, section 5.5.2.
 type PublicKey struct {
+	version      int
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/public_key.go:38]: This was my initial approach, however:
- I think having internal/unexported `version` is better API design, since the version gets set internally, and allows to set a default in only one place the `Version()` function. Moreover, this prevents version downgrades on keys/signatures that could lead to i.e. checking key Ids instead of fingerprints and losing properties of `v5` keys. Instead, the user can't modify the version (only input keys/signatures ofc), and if the code runs into an odd version (i.e., 0 - not set), it just defaults to whatever (currently, 4) and tries with that, or errors.
- There are some changes to make in the code. For example. `NewEntity` it needs to set e.g. version 4 (so that it is not 0). This involves setting the version in functions like (RSA) `GenerateKeyWithPrimes` or adding some logic before (`newSigner`).

It can be done but I largely prefer this design.

File: openpgp/key_generation.go [hunk 9]
```diff
@@ -34,6 +34,9 @@ func NewEntity(name, comment, email string, config *packet.Config) (*Entity, err
 		return nil, err
 	}
 	primary := packet.NewSignerPrivateKey(creationTime, primaryPrivRaw)
+	if config != nil && config.V5Keys {
+		primary.UpgradeToV5()
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/key_generation.go:38]: The responsibility of `UpgradeToV5` is not only setting the version, but also to `setFingerprintAndKeyId`. To remove these, one needs to call `setFingerprintAndKeyId` in 4 places.

File: openpgp/packet/signature.go (hunk 4 above)
Comment (zugzwang) [reply openpgp/packet/signature.go:846]: (To improve readability (and explicit initialization, not appending to nil))

File: openpgp/packet/signature.go (hunk 5 above)
Comment (zugzwang) [reply openpgp/packet/signature.go:938]: Maybe this was `go fmt`, not sure. But the above is unreadable to me (on the very least, break the line after `{`?)

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/public_key.go:38]: Yeah, but I think having `version` as `0` and defaulting to `4` is kind of hacky; I would rather have `NewEntity` set the version explicitly, anyway. And then, instead of setting it to 4 and calling `UpgradeToV5()`, it could just directly set it to the version that it wants.

Code that has access to the packet object can already do lots of strange things if it wants, so I don't think that the fact that it can modify the packet version is a huge issue. There are lots of other ways it could break the key / break security.

File: openpgp/packet/signature.go (hunk 4 above)
Comment (twiss) [reply openpgp/packet/signature.go:844]: But the value is already initialized as an out parameter, to the nil slice, which in golang kinda acts as an empty slice, and it's legal to append to it. Initializing it twice seems a bit redundant to me.

Comment (zugzwang) [reply openpgp/packet/signature.go:844]: Yes, I know it is legal to append to the nil slice. Indeed, I missed that it is already initialized in the function signature

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/public_key.go:38]: I'll export Version then, although don't like contributing further to strange behavior. Also, I'll need to modify every test initializing a public key or signature in order to include the version parameter 
(this is what I meant by the `0` version; an application setting e.g. 
```
pk := &PublicKey{<values>}
```
if it does not set `Version` then we should assume there's a default instead of failing with `public key version 0` somewhere).

[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/public_key.go:38]: (also, we need to modify all of `New{RSA, DSA, ElGamal, ECDSA, ECDH, EdDSA}PublicKey`, since they're part of the API they need to output keys with the correct `PublicKey.Version`)

File: openpgp/key_generation.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/key_generation.go:38]: (and `setFingerprintAndKeyId` is not exported so it can't be called there, actually)

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/public_key.go]: But also when they do `&PublicKey`, they would have to still call `setFingerprintAndKeyId`, right? So creating the packet object manually is a bit useless, and everyone should already be using `NewRSAPublicKey`. And then for `New{RSA, DSA, ElGamal, ECDSA, ECDH, EdDSA}PublicKey`; yeah, it is a bit unfortunate that they don't have a `config` parameter. Maybe for now, we could just set `Version: 4` in all of them, and then upgrade to v5 in `NewEntity` after all, indeed. Even though I think it's a bit ugly, but people that call `NewRSAPublicKey` might expect a v4 key, anyway.

File: openpgp/key_generation.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/key_generation.go:38]: OK, fair enough :+1: Yeah, let's keep `UpgradeToV5`, then

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/public_key.go]: Yes, you cannot call `setFingerprintAndKeyId` from the outside, but you can always set the fingerprint and key ID directly. If some user is doing so, this change is breaking (they'll need to put `Version: 4`).

File: openpgp/packet/signature.go (hunk 3 above)
Comment (zugzwang) [reply openpgp/packet/signature.go]: For V5 they MUST be equal and for V4 they SHOULD be equal, this is complying with that. `Signature` has already 35 fields or so, I'd prefer not to add more

File: openpgp/packet/signature.go [hunk 10]
```diff
@@ -58,6 +59,7 @@ type Signature struct {
 	PreferredSymmetric, PreferredHash, PreferredCompression []uint8
 	PreferredAEAD                                           []uint8
 	IssuerKeyId                                             *uint64
+	IssuerKeyFingerprint                                    []uint8
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/signature.go:62]: Isn't `uint8` an alias for`byte`? 2**8 = 256

File: openpgp/packet/signature.go [hunk 11]
```diff
@@ -443,6 +455,16 @@ func subpacketLengthLength(length int) int {
...
 }
 
+func (sig *Signature) CheckKeyIdOrFingerprint(pk *PublicKey) bool {
+	if sig.IssuerKeyId != nil && *sig.IssuerKeyId == pk.KeyId {
+		return true
+	}
+	if sig.IssuerKeyFingerprint == nil {
+		return false
+	}
+	return bytes.Equal(pk.Fingerprint, sig.IssuerKeyFingerprint)
```
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/signature.go]: :+1:

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/public_key.go]: Yeah, but.. I think it's fair to say that the API as is expects you to call `NewRSAPublicKey` etc, and that someone who isn't doing so is just using the API wrong :sweat_smile:

File: openpgp/packet/signature.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/signature.go]: Yeah, functionally they're the same, but there's a semantic difference. `uint8` is a number that fits in 8 bits, `byte` is a unit of data.

File: openpgp/packet/signature.go (hunk 3 above)
Comment (twiss) [reply openpgp/packet/signature.go]: Yeah, I'm not saying we should change that behavior, I'm just saying we should make it clear what the spec says by making the version come from the key instead of the signature.

You could also pass the signing key to this function, if you want.

File: openpgp/packet/public_key.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/public_key.go]: Right, still an API breaking change to consider (maybe someone who reads the values from somewhere else (say, not in armored format) and doesn't need to generate a new key?); the new `Version` field becomes compulsory.

File: openpgp/packet/signature.go (hunk 1 above)
Comment (zugzwang) [reply openpgp/packet/signature.go:751]: This should work now, given that the length is consistent, it only takes into account the 4 octet v/s 8 octet difference.

File: openpgp/packet/signature.go (hunk 2 above)
Comment (zugzwang) [reply openpgp/packet/signature.go:769]: This would involve re-serializing all subpackets (via `serializeSubpackets(., ., true)`, which in turns needs a byte array of the correct length), while this is already done and stored in `HashSuffix`. I prefer to have it done once, (less operations, less errors).

Comment (zugzwang) [reply openpgp/packet/signature.go]: It is possible though to reparse the hashed subpacket length and still reuse the `HashSuffix` (see last commit).

File: openpgp/packet/signature.go [hunk 12]
```diff
@@ -505,38 +529,52 @@ func (sig *Signature) SigExpired(currentTime time.Time) bool {
...
+		uint8(sig.Version),
+		uint8(sig.SigType),
+		uint8(sig.PubKeyAlgo),
+		uint8(hash),
+		uint8(len(hashedSubpackets) >> 8),
+		uint8(len(hashedSubpackets)),
+	})
+	hashedFields.Write(hashedSubpackets)
+
+	l := 6 + len(hashedSubpackets)
```
Comment (wussler) [inline openpgp/packet/signature.go:549]: `l` is an int, that can be 32 or 64 bits. I think here we want to force an uint64 type, if then we want to shift it by >32 bits

File: openpgp/packet/signature.go [hunk 13]
```diff
@@ -552,15 +590,16 @@ func (sig *Signature) Sign(h hash.Hash, priv *PrivateKey, config *Config) (err e
 	if priv.Dummy() {
 		return errors.ErrDummyPrivateKey("dummy key found")
 	}
-	sig.outSubpackets, err = sig.buildSubpackets()
+	sig.Version = priv.PublicKey.Version
+	sig.IssuerFingerprint = priv.PublicKey.Fingerprint
```
Comment (wussler) [inline openpgp/packet/signature.go:594]: Does this get set on v4 too?

File: openpgp/packet/signature.go [hunk 14]
```diff
@@ -719,12 +760,14 @@ func (sig *Signature) Serialize(w io.Writer) (err error) {
 }
 
 func (sig *Signature) serializeBody(w io.Writer) (err error) {
-	unhashedSubpacketsLen := subpacketsLength(sig.outSubpackets, false)
-	_, err = w.Write(sig.HashSuffix[:len(sig.HashSuffix)-6])
+	hashedSubpacketsLen := int(sig.HashSuffix[4])<<8 | int(sig.HashSuffix[5])
```
Comment (wussler) [inline openpgp/packet/signature.go:763]: I'd use unsigned int, since it's a length

File: openpgp/packet/signature.go [hunk 15]
```diff
@@ -771,17 +814,20 @@ type outputSubpacket struct {
...
 	subpackets = append(subpackets, outputSubpacket{true, creationTimeSubpacket, false, creationTime})
 
-	if sig.IssuerKeyId != nil {
+	if sig.IssuerKeyId != nil && sig.Version == 4 {
 		keyId := make([]byte, 8)
 		binary.BigEndian.PutUint64(keyId, *sig.IssuerKeyId)
-		subpackets = append(subpackets, outputSubpacket{true, issuerSubpacket, false, keyId})
+		subpackets = append(subpackets, outputSubpacket{true, issuerSubpacket, true, keyId})
+	}
+	if sig.IssuerFingerprint != nil {
```
Comment (wussler) [inline openpgp/packet/signature.go:827]: (see previous comment) if we set `IssuerFingerprint` for v4 too, we might end up serializing it too, but maybe I missed some logic

File: openpgp/write.go [hunk 16]
```diff
@@ -435,27 +443,30 @@ type signatureWriter struct {
 	signer        *packet.PrivateKey
 	sigType       packet.SignatureType
 	config        *packet.Config
+	metadata      *packet.LiteralData // Protected by v5 keys only
```
Comment (wussler) [inline openpgp/write.go:446]: Protected -> Produced?

File: openpgp/write.go [hunk 17]
```diff
@@ -435,27 +443,30 @@ type signatureWriter struct {
...
 
 func (s signatureWriter) Close() error {
 	sig := &packet.Signature{
+		Version:      s.signer.Version,
 		SigType:      s.sigType,
 		PubKeyAlgo:   s.signer.PubKeyAlgo,
 		Hash:         s.hashType,
 		CreationTime: s.config.Now(),
 		IssuerKeyId:  &s.signer.KeyId,
+		Metadata:      s.metadata,
```
Comment (wussler) [inline openpgp/write.go:469]: The final nitpick, not even @twiss can beat me: there is a space too much :joy:

**Review by wussler — COMMENTED** [review]
Well done @zugzwang!
The PR looks overall clean, tested, and straightforward.
I left some small comments, mostly about types and a couple of typos.

Cheers

File: openpgp/packet/private_key.go [hunk 18]
```diff
@@ -228,95 +253,107 @@ func mod64kHash(d []byte) uint16 {
...
 	}
-	privateKeyBytes := buf.Bytes()
+	l := priv.Len() // Private material count
 	if pk.sha1Checksum {
 		h := sha1.New()
-		h.Write(privateKeyBytes)
-		sum := h.Sum(nil)
-		privateKeyBytes = append(privateKeyBytes, sum...)
+		io.Copy(h, priv)
+		priv.Write(h.Sum(nil))
```
[the remark this answers is not in the document]
Comment (wussler) [reply openpgp/packet/private_key.go:347]: Sounds good to me

File: openpgp/packet/signature.go (hunk 15 above)
Comment (wussler) [reply openpgp/packet/signature.go:827]: I thought this was a v5 feat, glad to know it's right!

File: openpgp/write.go (hunk 16 above)
Comment (wussler) [reply openpgp/write.go:446]: I still don't understand the comment. :/
Is this available only in v5 keys?
Should this be protected only in v5 keys?

File: openpgp/write.go (hunk 17 above)
Comment (wussler) [reply openpgp/write.go]: Here, and indulgence to grant you pardon:
https://upload.wikimedia.org/wikipedia/commons/thumb/4/4e/Indulgence_San_Giovanni_in_Laterano_2006-09-07.jpg/1920px-Indulgence_San_Giovanni_in_Laterano_2006-09-07.jpg

**Comment by wussler** [comment]
Thanks for addressing the comments @zugzwang. LGTM now, but I'll leave the pleasure to approve to @twiss :)

File: openpgp/write.go (hunk 16 above)
Comment (zugzwang) [reply openpgp/write.go:446]: Then the comment is definitely misleading, I will change it.
This metadata (format, filename, time) is hashed by v5 signatures, therefore protecting the integrity. In contrast, in a v4 signature, I am free to change the filename or creation time within the literal packet before the signature, and it will still validate against the sig.

File: openpgp/packet/private_key.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/private_key.go:347]: `buf.Bytes()` doesn't actually allocate any memory, it just returns a slice pointing to the bytes.

File: openpgp/packet/private_key.go [hunk 19]
```diff
@@ -423,7 +460,6 @@ func (pk *PrivateKey) Encrypt(passphrase []byte) error {
 	}
 
 	pk.s2kParams, err = s2k.Generate(rand.Reader, s2kConfig)
-	privateKeyBytes := privateKeyBuf.Bytes()
```
Comment (twiss) [inline openpgp/packet/private_key.go:426]: Also here I prefer the old code of only calling `privateKeyBuf.Bytes()` once, instead of multiple times

File: openpgp/key_generation.go [hunk 20]
```diff
@@ -34,20 +34,28 @@ func NewEntity(name, comment, email string, config *packet.Config) (*Entity, err
...
+	} else {
+		primary.Version = 4
+	}
```
Comment (twiss) [inline openpgp/key_generation.go:41]: Is this needed, since you set the version to 4 in all the low-level functions?

Comment (zugzwang) [reply openpgp/key_generation.go:41]: Not anymore :+1:

File: openpgp/packet/private_key.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/private_key.go:347]: True, I was mixing up with the previous code where memory do gets allocated in `privateKeyBytes`. Anyway I'd prefer to avoid copying those private bytes too much around.

[the remark this answers is not in the document]
Comment (twiss) [reply openpgp/packet/private_key.go:347]: But `privateKeyBytes := buf.Bytes()` also doesn't copy the data, it just assigns the slice which contains a pointer to the data.

[the remark this answers is not in the document]
Comment (zugzwang) [reply openpgp/packet/private_key.go:347]: Ah, my bad then.

**Review by twiss — APPROVED** [review]
